A42-xds-ring-hash-lb-policy.md

```diff
@@ -0,0 +1,587 @@
+additional policy, `RING_HASH`.
+
+The configuration for the Ring Hash LB policy is in the
+[`ring_hash_lb_config`
+field](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L907).
+The field is optional; if not present, defaults will be assumed for all
+of its values.  gRPC will support the
+[`minimum_ring_size`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L397)
+and
+[`maximum_ring_size`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L406)
+fields the same way that Envoy does.  The
+[`hash_function`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L401)
+field will be required to be set to `XX_HASH`; if it is set to any other
+value (at present, the only other value is `MURMUR_HASH_2`), gRPC will
+NACK the `Cluster` resource.
+
+#### Background: Envoy vs. gRPC LB Policies
+
+Envoy's LB policies are expected to handle both locality-picking and
+endpoint-picking in a single layer.  It does not currently have a clean
+way to independently select the locality-picking and endpoint-picking
+policies the way that we do in gRPC.  For example, with the `ROUND_ROBIN`
+LB policy, Envoy has a single policy that handles both picking the
+locality and the endpoint within that locality as a single step.
+In contrast, in gRPC, we implement the xDS `ROUND_ROBIN` policy by
+having a `weighted_target` LB policy that chooses the locality and then
+delegates to a nested `round_robin` policy for that locality to choose
+the endpoint within the locality.
+
+Unfortunately, we cannot do the same thing with the `RING_HASH` policy.
+The `RING_HASH` LB policy in Envoy handles both locality-picking and
+endpoint-picking in a single layer.  It basically puts all endpoints in
+all localities into a single ring and then chooses an endpoint for a
+request from the ring based on the request's hash.  We cannot split this
+up into two different policies, because the two policies would pick
+independently, which means that the failover behavior would cause
+unnecessary thrashing in the choice of endpoint when the originally
+chosen endpoint is down.  For example, consider the case where the
+endpoint nearest to the request's hash is in locality A and the
+next-nearest is in locality B.  If the client sees that the endpoint
+in locality A is unreachable before it sees an EDS update that removes
+that endpoint, the endpoint-picking policy for locality A would switch
+to the next-closest endpoint in the same locality.  Then, when the EDS
+update removes that endpoint from the list, the locality-picking endpoint
+would choose locality B instead.  This would result in switching backends
+twice instead of just once, which is too much churn.  As a result, gRPC
```

Comment:
Mixing locality-level load balancing weight with endpoint-level load balancing weight may actually give incorrect balancing behavior. The intra-locality endpoint-level load balancing weight is [optional](https://github.com/envoyproxy/envoy/blob/587ae281a6d2e67a338a7caa5ee6d7c240a604b5/api/envoy/config/endpoint/v3/endpoint_components.proto#L105-L107). If it is unspecified, endpoints _within the locality_ should be balanced equally. Note, that behavior should be limited to intra-locality load balancing only. If we mix the locality-level weight and endpoint-level weight by simply multiplying them together, the normalized balancing weight for each host may be incorrect. Think about the following example:
- Locality A with weight of 100, containing endpoint1 and endpoint2. But endpoints in this locality do not have intra-locality endpoint-level weight specified.
- Locality B with weight of 1, containing endpoint3 and endpoint4. Each endpoint has intra-locality endpoint-level weight being 100.

When mixing the two levels of weight together, the effective balancing behavior would be round robin of all the 4 endpoints. But this is not really the configuration intends to mean: endpoint1 and endpoint2 should receive significantly larger amount of RPCs than endpoint3 and endpoint4.

So the real problem is the locality-level weight and the intra-locality endpoint-level weight are targeting on two different scopes of things. They cannot be simply multiplied together. Things can only work well if all localities have the same normalization of endpoint-level weights. That is, we'd require either all endpoints in all localities not specifying their intra-locality weight or all endpoints in all localities must set their intra-locality weight. We may want to mention this in user guide maybe?

Reply:
I believe that this multiplication is exactly what Envoy does.  I think in the scenario you describe, giving all four endpoints equal weight is exactly the right behavior.